llvm/lib/Target/X86/X86ISelLoweringCall.cpp

```diff
@@ -2723,11 +2722,19 @@ bool MatchingStackOffset(SDValue Arg, unsigned Offset, ISD::ArgFlagsTy Flags,
 
 /// Check whether the call is eligible for tail call optimization. Targets
 /// that want to do tail call optimization should implement this function.
+/// Note that musttail calls are not checked for eligibility, so the backend
+/// must support forwarding arguments of any type.
```

Comment:
This doesn't really get the relevant point across... I would say:

This function is also used to reject musttail calls the backend does not know how to emit.  musttail calls should be supported for all combinations of argument/return type.

Reply:
What you said is true for the ARM backend, but this is the X86 backend, where the opposite is true, so I think this is correct as written. This was a point of disagreement in the past, how to approach musttail call eligibility checks.

The ARM backend is extremely conservative, resulting in rejecting tail calls we can probably support (sret forwarding). The X86 backend aggressively assumes it can support everything, because it already does in order to support GuaranteedTCO (not supported by ARM). In practice, this does result in miscompilations (byval forwarding, https://github.com/llvm/llvm-project/issues/46402), so there is a tradeoff between optimistic and conservative.

Reply:
Oh, I see... in that case, maybe just say that the x86 backend specifically doesn't call this for musttail calls.

(I generally support reporting errors over miscompiling; there are probably arguments either way for structuring this specific part of the code, though.)